lib/src/common/disposable.dart

```diff
@@ -331,13 +364,20 @@ class Disposable implements _Disposable, DisposableManagerV6, LeakFlagger {
     if (isDisposed) {
       return null;
     }
-    if (_isDisposing) {
+    if (isDisposalImminent || isDisposing) {
       return didDispose;
     }
-    _isDisposing = true;
 
-    await Future.wait(_awaitableFutures);
-    _awaitableFutures.clear();
+    _isDisposalImminent = true;
+
+    while (_awaitableFutures.isNotEmpty) {
+      final futures = _awaitableFutures.toList();
+      _awaitableFutures.clear();
+      await Future.wait(futures);
+    }
+
+    _isDisposing = true;
+    _isDisposalImminent = false;
```

Comment:
FWIW: Setting true/false flags for each state is easy to goof up. Is it better to have an enum mutually exclusive states and use that to set the current state of the disposable object?